File: backend/routers/vehicles.py
"""
AutoWorth AI — Vehicles Router
GET /api/v1/vehicles
GET /api/v1/vehicles/{id}
POST /api/v1/vehicles (admin only)
"""
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from ..database import get_db
from ..dependencies import get_current_active_user, get_admin_user
from ..models import User, Vehicle
from ..schemas import VehicleResponse, VehicleCreate, PaginatedResponse

router = APIRouter(prefix="/vehicles", tags=["Vehicles"])


@router.get("", response_model=PaginatedResponse)
def list_vehicles(
    brand: Optional[str] = Query(None),
    model: Optional[str] = Query(None),
    fuel_type: Optional[str] = Query(None),
    transmission: Optional[str] = Query(None),
    year_min: Optional[int] = Query(None),
    year_max: Optional[int] = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    _: User = Depends(get_current_active_user),
):
    query = db.query(Vehicle)
    if brand:
        query = query.filter(Vehicle.brand.ilike(f"%{brand}%"))
    if model:
        query = query.filter(Vehicle.model.ilike(f"%{model}%"))
    if fuel_type:
        query = query.filter(Vehicle.fuel_type == fuel_type)
    if transmission:
        query = query.filter(Vehicle.transmission == transmission)
    if year_min:
        query = query.filter(Vehicle.year >= year_min)
    if year_max:
        query = query.filter(Vehicle.year <= year_max)

    total = query.count()
    items = query.offset((page - 1) * page_size).limit(page_size).all()
    return PaginatedResponse(
        items=[VehicleResponse.model_validate(v) for v in items],
        total=total,
        page=page,
        page_size=page_size,
        pages=(total + page_size - 1) // page_size,
    )


@router.get("/{vehicle_id}", response_model=VehicleResponse)
def get_vehicle(
    vehicle_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_active_user),
):
    vehicle = db.query(Vehicle).filter(Vehicle.id == vehicle_id).first()
    if not vehicle:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Vehicle not found")
    return vehicle


@router.post("", response_model=VehicleResponse, status_code=status.HTTP_201_CREATED)
def create_vehicle(
    payload: VehicleCreate,
    db: Session = Depends(get_db),
    admin: User = Depends(get_admin_user),
):
    vehicle = Vehicle(**payload.model_dump())
    db.add(vehicle)
    db.commit()
    db.refresh(vehicle)
    return vehicle

File: frontend/src/api/vehicles.ts
import client from './client';
import type { Vehicle, PaginatedResponse } from '../types';

export const vehiclesApi = {
  /** List vehicles with optional filters */
  list: (params?: {
    brand?: string;
    model?: string;
    fuel_type?: string;
    year_min?: number;
    year_max?: number;
    page?: number;
    page_size?: number;
  }) =>
    client.get<PaginatedResponse<Vehicle>>('/vehicles', { params }).then((r) => r.data),

  /** Get a single vehicle by ID */
  get: (id: number) =>
    client.get<Vehicle>(`/vehicles/${id}`).then((r) => r.data),

  /** Get distinct brands available in the database */
  getBrands: () =>
    client.get<string[]>('/vehicles/brands').then((r) => r.data),

  /** Get models for a given brand */
  getModels: (brand: string) =>
    client.get<string[]>('/vehicles/models', { params: { brand } }).then((r) => r.data),

  /** Get all available cities */
  getCities: () =>
    client.get<string[]>('/vehicles/cities').then((r) => r.data),

  /** Save a vehicle to the user's saved cars */
  save: (vehicleId: number, savedPrice?: number) =>
    client
      .post('/saved-cars', { vehicle_id: vehicleId, saved_price: savedPrice })
      .then((r) => r.data),

  /** Remove a vehicle from saved cars */
  unsave: (vehicleId: number) =>
    client.delete(`/saved-cars/${vehicleId}`).then((r) => r.data),

  /** List saved cars for the authenticated user */
  listSaved: () =>
    client.get('/saved-cars').then((r) => r.data),
};

export default vehiclesApi;
